core/embed/rust/src/micropython/dict.rs

```diff
@@ -0,0 +1,68 @@
+use core::convert::TryFrom;
+
+use crate::error::Error;
+
+use super::{ffi, gc::Gc, map::Map, obj::Obj};
+
+/// Insides of the MicroPython `dict` object.
+pub type Dict = ffi::mp_obj_dict_t;
+
+impl Dict {
+    /// Allocate a new dictionary on the GC heap, empty, but with a capacity of
+    /// `capacity` items.
+    pub fn alloc_with_capacity(capacity: usize) -> Gc<Self> {
+        unsafe {
+            let ptr = ffi::mp_obj_new_dict(capacity);
+            Gc::from_raw(ptr.as_ptr().cast())
+        }
+    }
+
+    /// Constructs a dictionary definition by taking ownership of given [`Map`].
+    pub fn with_map(map: Map) -> Self {
+        unsafe {
+            Self {
+                base: ffi::mp_obj_base_t {
+                    type_: &ffi::mp_type_dict,
+                },
+                map,
+            }
+        }
+    }
+
+    /// Returns a reference to the contained [`Map`].
+    pub fn map(&self) -> &Map {
+        &self.map
+    }
+
+    /// Returns a mutable reference to the contained [`Map`].
+    pub fn map_mut(&mut self) -> &mut Map {
+        &mut self.map
+    }
+}
+
+impl From<Gc<Dict>> for Obj {
+    fn from(value: Gc<Dict>) -> Self {
+        // SAFETY:
+        //  - `value` is an object struct with a base and a type.
+        //  - `value` is GC-allocated.
```

Comment:
> These safety notes confuse me more than help. Aren't both obvious from the type of the value?

I'll reword these.

> esp. the "gc-allocated" thing, you're calling Gc::into_raw but I'd expect value.into_raw() (not sure if that is illegal or impossible or?)

It's best to avoid implementing methods for the smart pointer types, because they `Deref` into the inner type and it's not immeadiately obvious which type the method belongs to.